### PR TITLE
5756 implement bearer token authentication for external api clients

### LINKED ISSUE #5756: Implement bearer token authentication for external API clients
## Background ##

Enable external tools (Postman, CLI tools, CI/CD) to authenticate against the Django API using Keycloak-issued bearer tokens, without requiring browser-based session cookies. This is a new capability that the legacy auth architecture cannot support easily.

Parent epic: #5703
Architecture plan: `docs/Technical-Documentation/tech-memos/keycloak/keycloak-architecture-plan.md` (Section 7 — External Tool and API Client Authentication)
Depends on: #5757 (canary at 100%)

### Why It Matters ###

The legacy auth flow is tightly coupled to the browser — external tools must reverse-engineer session cookies to call the API. Keycloak enables standard OAuth2 grants (Authorization Code + PKCE, Device Authorization) that work with Postman, CLI tools, and CI/CD pipelines out of the box. This also simplifies access for security auditors and penetration testers.

### What's Changing ###

Django gains a bearer token authentication path alongside session-based auth. New public Keycloak clients (`tdp-cli`) allow external tools to authenticate via standard OAuth2 flows — no client secrets to distribute.

### Acceptance Criteria ###

**Django Bearer Token Support**
- [ ] Add middleware or DRF authentication backend that validates Keycloak-issued JWT bearer tokens against the JWKS endpoint
- [ ] Resolve Django user from token claims (`login_gov_uuid`, `hhs_id`, email) — same logic as `KeycloakOIDCBackend`
- [ ] Ensure Django authorization (permissions, STT scoping, approval status) works identically for token-authenticated requests
- [ ] Add per-client rate limiting for API token requests
- [ ] Ensure token-authenticated requests are logged with client ID for audit

**Keycloak Client Registration**
- [ ] Register `tdp-cli` as a **public** Keycloak client (no client secret) with Authorization Code + PKCE and Device Authorization Grant
- [ ] Scope clients to minimum required grants and scopes (principle of least privilege)

**Validation**
- [ ] Test Postman OAuth2 flow: configure with Keycloak discovery URL, `tdp-cli` client ID, PKCE S256 — authenticate via Login.gov/AMS in popup, receive bearer token, call Django API
- [ ] Test Device Authorization Grant: CLI displays URL + code, user authenticates in browser, CLI receives token
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

### Notes ###
- This enables: CLI-based STT data submission, scripted report generation, and simplified access for security auditors/penetration testers.
- `mozilla-django-oidc` supports bearer token validation via `OIDCAuthentication` for DRF, or a lightweight custom middleware can be used.
- Users still authenticate through Login.gov/AMS — the only difference is how the token is delivered to the requesting tool.

### Supporting Documentation ###
- Architecture plan: Section 7 — External Tool and API Client Authentication
- `mozilla-django-oidc` DRF integration docs

### Open Questions ###
- None currently

### PR DESCRIPTION
**Pull request by mattcoleanderson**
## Summary of Changes

Pull request closes #5756.

Adds JWT bearer-token authentication so external API clients (Postman, CLI tools, CI/CD, security auditors) can authenticate against the Django API using standard OAuth2 grants. 

Users still authenticate through Login.gov / AMS via Keycloak; only the token-delivery mechanism changes.

### Setup

The `tdp-cli` client is baked into `realm-export.json`, so a local Keycloak that already has its DB populated will not see it until the realm is re-imported. To pick it up:

```bash
cd tdrs-backend
task backend-down
docker volume rm tdrs-backend_keycloak_pg_data
docker compose build --no-cache keycloak
task backend-update    # rebuilds backend image so mozilla-django-oidc is installed
task backend-up
```

In Keycloak admin (http://localhost:8443, `admin`/`admin`) → realm `tdp` → **Clients**, confirm `tdp-cli` is listed alongside `tdp-django` and `tdp-grafana`.

### Create a Keycloak test user

The Django `KeycloakSyncClient` updates existing Keycloak users but doesn't create them. That creation normally happens via the Login.gov / AMS first-broker-login flow. For local testing, its easiest to drop your user from the DB, then go through the new user login flow to auto create a user in Keycloak. Then you can add a credential in Keycloak:

1. Keycloak admin → realm `tdp` → **Users** → your user → **Credentials** tab → **Set password**. Temporary OFF. Save.

### Postman (Authorization Code + PKCE)
In postman create a new request then open the Authorization tab, select Type **OAuth 2.0** then click: **Configure New Token**.

Now input the following values:

| Field | Value |
|---|---|
| Grant Type | **Authorization Code (With PKCE)** |
| Callback URL | `https://oauth.pstmn.io/v1/callback` |
| Auth URL | `http://localhost:8443/realms/tdp/protocol/openid-connect/auth` |
| Access Token URL | `http://localhost:8443/realms/tdp/protocol/openid-connect/token` |
| Client ID | `tdp-cli` |
| Client Secret | *(empty)* |
| Code Challenge Method | **SHA-256** |
| Scope | `openid email profile tdp-user-attributes` |
| Client Authentication | **Send client credentials in body** |

Click **Get New Access Token** → log in with the Keycloak user → token returned → **Use Token**.

Then:
- `GET http://localhost:8080/v1/users/` with the bearer token → expect **200** and a paginated user list.
- Same request with the `Authorization` header dropped → expect **401**.
- `task backend-logs | grep "Bearer token auth"` shows a log line tagged `client=tdp-cli user=<email> path=/v1/users/`.

### CLI (Device Authorization Grant) via curl

```bash
export CODE_VERIFIER=$(openssl rand -base64 64 | tr -d "=\n" | tr "+/" "-_" | head -c 64)
export CODE_CHALLENGE=$(printf "%s" "$CODE_VERIFIER" | openssl dgst -sha256 -binary | openssl base64 | tr -d "=\n" | tr "+/" "-_")

curl -s -X POST "http://localhost:8443/realms/tdp/protocol/openid-connect/auth/device" \
    -d "client_id=tdp-cli" \
    -d "scope=openid email profile tdp-user-attributes" \
    -d "code_challenge=$CODE_CHALLENGE" \
    -d "code_challenge_method=S256" | jq
```

Open `verification_uri_complete` from the response in a browser, log in, **Grant access**. Then:

```bash
DEVICE_CODE="<paste device_code from previous response>"
curl -s -X POST "http://localhost:8443/realms/tdp/protocol/openid-connect/token" \
    -d "grant_type=urn:ietf:params:oauth:grant-type:device_code" \
    -d "device_code=$DEVICE_CODE" \
    -d "client_id=tdp-cli" \
    -d "code_verifier=$CODE_VERIFIER" | jq
```

Returns the access token. Use it:

```bash
TOKEN="<paste access_token>"
curl -s -H "Authorization: Bearer $TOKEN" http://localhost:8080/v1/users/ | jq
```

Expect the same 200 + JSON, with another `Bearer token auth client=tdp-cli` audit log line. Same client, different grant, identical authorization.

### Authorization parity

Hit an STT-scoped endpoint — `GET http://localhost:8080/v1/data_files/` — with a bearer token from a non-admin user and confirm the response is filtered the same way as in the browser session. (Permission/STT logic runs identically because the bearer authenticator returns the same `request.user` as the OIDC backend would.)

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] DRF authentication backend validates Keycloak-issued JWT bearer tokens against the JWKS endpoint
+ [x] Django user resolved from token claims (`login_gov_uuid`, `hhs_id`, email) — same logic as `KeycloakOIDCBackend`
+ [x] Authorization (permissions, STT scoping, approval status) works identically for token-authenticated requests
+ [x] Per-client rate limiting for API token requests (`KEYCLOAK_CLIENT_RATE`, default 300/min)
+ [x] Token-authenticated requests logged with client ID for audit
+ [x] `tdp-cli` registered as a **public** Keycloak client (no secret) with Authorization Code + PKCE and Device Authorization Grant
+ [x] Postman OAuth2 flow validated end-to-end
+ [x] Device Authorization Grant validated end-to-end
+ [ ] **\`lfrohlich\`** and/or **\`adpennington\`** confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested? *(no frontend changes)*
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see \`CodeCov Report\` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see \`CodeCov Report\` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with \`iamjolly\` and \`ttran-hub\` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Review by elipe17 — CHANGES_REQUESTED** [review]
The login.gov auth option has `"hideOnLogin": true` in the realm-export.json. This is an artifact of my MVP branch and was configured that way to initially prevent login.gov authenticated users from using SSO with PLG. Since that initial step, the auth flow in Grafana has been set up to track the user group which prevents any non developer and system admin from authenticating regardless of their IdP. Unhiding login.gov from the realm export will allow us to authenticate with that IdP for the `tdp-cli` client instead of having to assign credentials in the keycloak console and no longer pose a threat to exposing Grafana to unprivileged users.

**Comment by mattcoleanderson** [comment]
> The login.gov auth option has `"hideOnLogin": true` in the realm-export.json. This is an artifact of my MVP branch and was configured that way to initially prevent login.gov authenticated users from using SSO with PLG. Since that initial step, the auth flow in Grafana has been set up to track the user group which prevents any non developer and system admin from authenticating regardless of their IdP. Unhiding login.gov from the realm export will allow us to authenticate with that IdP for the `tdp-cli` client instead of having to assign credentials in the keycloak console and no longer pose a threat to exposing Grafana to unprivileged users.

@elipe17 Nice! Much better than creating a credential for individual users. I've removed `"hideOnLogin": true` and confirmed that users can now authenticate with Login.gov sign-in through the oauth flow. No need to create the credential anymore and also, now users are added directly to Keycloak when logging into Login.gov

**Review by elipe17 — APPROVED** [review]
This is looking great and working as expected. One thing I think that would be cool to investigate a little is aligning Keycloak's login form with TDPs. Keycloak has the ability to customize and them pretty much the entire app. It would be nice to try and get that closer to 508 compliance since it will be a user facing page once we open the tdp-cli client to users.

File: tdrs-backend/tdpservice/users/authentication.py [hunk 1]
```diff
@@ -3,12 +3,115 @@
...
+        token = _bearer_token_from_request(request)
+        if not token:
+            return None
+
+        # OIDCAuthenticationBackend.verify_token retrieves JWKS, verifies the
+        # RS256 signature and the ``exp`` claim. Access tokens have no nonce,
+        # so the nonce branch is a no-op (None == None).
+        backend = OIDCAuthenticationBackend()
+        try:
+            payload = backend.verify_token(token)
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/users/authentication.py]: This reuses `OIDCAuthenticationBackend.verify_token()` for bearer access tokens, but that backend is configured with `OIDC_RP_CLIENT_ID = tdp-django` in common.py, while this feature expects tokens for the public `tdp-cli` client. If `mozilla-django-oidc` enforces audience/client-id matching, valid `tdp-cli` tokens will be rejected here; if it does not, then we are accepting tokens without proving they were issued for this API. We should use a verifier that explicitly validates the expected audience/authorized party for the external-client flow, and add a non-mocked test that would fail on a `tdp-cli` vs `tdp-django` mismatch.

**Review by jtimpe — COMMENTED** [review]
Looks great! I think Mo's comment could use additional consideration, once you've responded to that, re-request my review and I'm happy to approve

File: tdrs-backend/tdpservice/users/authentication.py (hunk 1 above)
Comment (mattcoleanderson) [reply tdrs-backend/tdpservice/users/authentication.py]: @raftmsohani good catch. I agree with your concern.

I updated this so bearer-token auth no longer relies only on the existing OIDC token verifier. The API now explicitly validates that the token is intended for this external-client flow by checking:

- `azp == tdp-cli`
- `aud` includes `tdp-django`
- `iss` matches the configured Keycloak realm issuer
- `exp` is valid

I also added signed JWT tests with mocked JWKS endpoint for the valid path and negative cases covering wrong/missing `azp`, wrong/missing `aud`, wrong issuer, and expired token.

I also updated the local Keycloak realm export to add an audience client scope for `tdp-django` and attach it only to `tdp-cli`, so CLI-issued tokens include the expected API audience.

File: tdrs-backend/tdpservice/users/authentication.py [hunk 2]
```diff
@@ -1,14 +1,179 @@
...
+        """Validate a bearer token and resolve the Django user, or return None."""
+        token = _bearer_token_from_request(request)
+        if not token:
+            return None
+
+        try:
+            payload = _verify_keycloak_bearer_token(token)
+        except (jwt.InvalidTokenError, requests.RequestException) as exc:
+            logger.info("Bearer token verification failed: %s", exc)
+            raise AuthenticationFailed(_("Invalid bearer token."))
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/users/authentication.py:151]: Nit: distinguish between expired vs malformed for better UX

Comment (mattcoleanderson) [reply tdrs-backend/tdpservice/users/authentication.py:151]: @raftmsohani Done

File: tdrs-backend/keycloak/realm-configs/realm-export.dev-local.json [hunk 3]
```diff
@@ -237,6 +237,28 @@
           }
         }
       ]
+    },
+    {
+      "name": "tdp-api-audience",
```
Comment (elipe17) [inline tdrs-backend/keycloak/realm-configs/realm-export.dev-local.json:242]: Ricky Bobby coming in hot here. Can we apply these realm changes to the staging and prod realms so we don't forget in the future?

Comment (mattcoleanderson) [reply tdrs-backend/keycloak/realm-configs/realm-export.dev-local.json:242]: @elipe17 Done!

File: tdrs-backend/keycloak/configure-idps.sh [hunk 4]
```diff
@@ -281,6 +289,143 @@ configure_master_realm_security_headers() {
     echo "Master realm X-Frame-Options set to SAMEORIGIN."
 }
 
+configure_tdp_client_urls() {
```
Comment (elipe17) [inline tdrs-backend/keycloak/configure-idps.sh:492]: I think this change is unsafe for the Keycloak deployment path.

`deploy.sh` does set `DEPLOY_ENV` in the app manifest, so staging/prod should not fall through to the `DEPLOY_ENV=dev` default. That means the main staging/prod risk is not localhost injection.

The issue is that `configure_tdp_client_urls` initializes `redirect_uris` and `web_origins` as empty arrays, then always writes those arrays back to the `tdp-django` client. The deploy task does not currently pass `KC_TDP_REDIRECT_URIS` or `KC_TDP_WEB_ORIGINS`, so staging/prod would overwrite the existing realm-export redirect URLs with `[]`, which would break browser login redirects.

For dev, the script does add localhost/127.0.0.1 URLs when `DEPLOY_ENV=dev`. Since the deploy path also does not pass cloud.gov frontend URLs, dev deployment likely ends up with only localhost redirect URIs/web origins for `tdp-django`, which would break deployed dev frontend login.

Suggested fix: either preserve the existing client `redirectUris`/`webOrigins` when the new vars are absent, or update `deploy.sh` to compute and pass the correct `KC_TDP_REDIRECT_URIS` and `KC_TDP_WEB_ORIGINS` for each environment before running `configure-idps.sh`.

Comment (mattcoleanderson) [reply tdrs-backend/keycloak/configure-idps.sh:492]: @elipe17 oh good catch. I updated `configure_tdp_clients_urls` to first fetch the current `tdp-django` client config, so now it will preserve the existing configuration when new vars are absent.

File: tdrs-backend/tdpservice/users/throttling.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    The rate is configured via DRF's ``DEFAULT_THROTTLE_RATES['keycloak_client']``
+    (see ``KEYCLOAK_CLIENT_RATE`` in settings). Counters live in the dedicated
+    Redis-backed ``throttle`` cache so they're shared across web workers.
+    """
+
+    cache = caches["throttle"]
+    scope = "keycloak_client"
+
+    def get_cache_key(self, request, view):
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/users/throttling.py:36]: The throttle key uses only client_id (the azp claim). Since the authenticator enforces a single tdp-cli client ID for all bearer tokens, all external API users share one global rate limit. A noisy script or compromised token will throttle all other bearer clients.

File: tdrs-backend/tdpservice/users/authentication.py [hunk 6]
```diff
@@ -1,14 +1,182 @@
...
+        settings.KEYCLOAK_DJANGO_CLIENT_ID,
+    )
+
+
+def _expected_keycloak_bearer_client_id():
+    """Return the Keycloak client allowed to use bearer auth."""
+    return getattr(settings, "KEYCLOAK_BEARER_CLIENT_ID", "tdp-cli")
+
+
+def _matching_jwks_key(token):
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/users/authentication.py:73]: JWKS is fetched on every bearer-token request with no caching -> suggestion is to add Cache JWKS by kid with TTL

Comment (elipe17) [reply tdrs-backend/tdpservice/users/authentication.py:73]: This is a good catch

File: tdrs-backend/tdpservice/users/throttling.py (hunk 5 above)
Comment (elipe17) [reply tdrs-backend/tdpservice/users/throttling.py:36]: Another good catch

**Comment by elipe17** [comment]
Will you try deploying this to the dev instance of keycloak and verify it successfully deploys and your realm/config changes stick?

File: tdrs-backend/tdpservice/users/throttling.py (hunk 5 above)
Comment (mattcoleanderson) [reply tdrs-backend/tdpservice/users/throttling.py:36]: @raftmsohani Just updated this so bearer-token throttling is keyed by both `client_id` and `user_id` instead of only `client_id`

File: tdrs-backend/tdpservice/users/authentication.py (hunk 6 above)
Comment (mattcoleanderson) [reply tdrs-backend/tdpservice/users/authentication.py:73]: @raftmsohani good suggestion, just pushed the change.

**Review by elipe17 — CHANGES_REQUESTED** [review]
Not seeing login.gov as an auth option when testing the deployed realm. Maybe we have to explicitly set login.gov as unhidden too?

Screenshot 2026-05-19 at 11 57 51 AM

**Comment by elipe17** [comment]
I am also unable to use the deployed instance to successfully query api endpoints in the raft environment without a "missing credentials" error.

**Comment by mattcoleanderson** [comment]
> Not seeing login.gov as an auth option when testing the deployed realm. Maybe we have to explicitly set login.gov as unhidden too?
> 
> Screenshot 2026-05-19 at 11 57 51 AM

@elipe17  Resolved. It was still hidden in the config.


> I am also unable to use the deployed instance to successfully query api endpoints in the raft environment without a "missing credentials" error.

@elipe17  I'm still looking down this rabbit hole. I've pushed what I believe will fix it, but need to build and deploy the app to raft env to test. I believe, ngnix wasn't forwarding the auth header to the backend. I've updated the locations.conf to do just that.

**Comment by mattcoleanderson** [comment]
> I am also unable to use the deployed instance to successfully query api endpoints in the raft environment without a "missing credentials" error.

image

@elipe17 fixed. You should be able to test this PR now in the raft env. I was right that nginx wasn't forwarding the Auth header

**Review by elipe17 — APPROVED** [review]
This is awesome sir! Working as expected. LGTM

**Review by jtimpe — APPROVED** [review]
Good work!
